src/player-attack.c

```diff
@@ -118,30 +140,83 @@ static int chance_of_missile_hit(const struct player *p,
 		chance = p->state.skills[SKILL_TO_HIT_BOW] + bonus * BTH_PLUS_ADJ;
 	}
 
-	return chance - distance(p->grid, grid);
+	return chance;
 }
 
 /**
- * Determine if the player "hits" a monster.
+ * Calculate the player's missile to-hit value against a specific monster.
+ * See also: chance_of_melee_hit
+ *
+ * \param p The player
+ * \param missile The missile to launch
+ * \param launcher Optional launcher to use (thrown weapons use no launcher)
+ * \param mon The monster
  */
-bool test_hit(int chance, int ac, int vis) {
-	int k = randint0(100);
+static int chance_of_missile_hit(const struct player *p,
+	const struct object *missile, const struct object *launcher,
+	const struct monster *mon)
+{
+	int chance = chance_of_missile_hit_base(p, missile, launcher);
+	/* Penalize for distance */
+	chance -= distance(p->grid, mon->grid);
+	/* Non-visible targets have a to-hit penalty of 50% */
+	return monster_is_obvious(mon) ? chance : chance / 2;
```

Comment:
Mimic change to use is_obvious for ranged attacks ends up here. I did a rebase so you can't see the result of the merge separately. Maybe you'd rather see a regular merge to resolve conflicts and do a squash rebase when pulling the branch in next time?